apiml/src/main/java/org/zowe/apiml/controller/ReactivePATController.java

```diff
@@ -0,0 +1,420 @@
+import org.zowe.apiml.security.common.token.TokenAuthentication;
+import org.zowe.apiml.util.HttpUtils;
+import org.zowe.apiml.zaas.controllers.AuthController.RulesRequestModel;
+import org.zowe.apiml.zaas.controllers.AuthController.ValidateRequestModel;
+import org.zowe.apiml.zaas.security.service.AuthenticationService;
+import org.zowe.apiml.zaas.security.service.TokenCreationService;
+import reactor.core.publisher.Mono;
+
+import java.io.IOException;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+
+import static org.apache.http.HttpStatus.SC_SERVICE_UNAVAILABLE;
+import static org.zowe.apiml.zaas.controllers.AuthController.ACCESS_TOKEN_REVOKE;
+import static org.zowe.apiml.zaas.controllers.AuthController.ACCESS_TOKEN_REVOKE_MULTIPLE;
+import static org.zowe.apiml.zaas.controllers.AuthController.ACCESS_TOKEN_VALIDATE;
+
+@RestController
+@RequestMapping("/gateway/api/v1/auth")
+@Slf4j
+@RequiredArgsConstructor
+public class ReactivePATController {
+
+    private static final String TOKEN_KEY = "token";
+
+    private final AccessTokenProvider tokenProvider;
+    private final RauditxService rauditxService;
+    private final AuthenticationService authenticationService;
+    private final TokenCreationService tokenCreationService;
+    private final MessageService messageService;
+    private final HttpUtils httpUtils;
+    private final ObjectMapper mapper;
+    private final PeerAwareInstanceRegistryImpl peerAwareInstanceRegistry;
+
+    @Data
+    @NoArgsConstructor
+    @AllArgsConstructor
+    public static class AccessTokenRequest {
+        private int validity;
+        private Set<String> scopes;
+    }
+
+    @PostMapping("/access-token/generate")
+    public Mono<ResponseEntity<String>> generatePat(@RequestBody AccessTokenRequest accessTokenRequest) {
+        return ReactiveSecurityContextHolder.getContext()
+            .map(SecurityContext::getAuthentication)
+            .filter(Objects::nonNull)
+            .<ResponseEntity<String>>handle((authentication, sink) -> {
+                if (accessTokenRequest.getScopes() == null || accessTokenRequest.getScopes().isEmpty()) {
+                    sink.error(new AccessTokenBodyNotValidException(AccessTokenBodyNotValidException.Reason.MISSING_SCOPES));
+                    return;
+                }
+
+                var userId = authentication.getName();
+
+                log.debug("Generating access token for user {}", userId);
+
+                RauditxService.RauditxBuilder rauditBuilder = rauditxService.builder()
+                    .userId(userId)
+                    .messageSegment("An attempt to generate PAT")
+                    .alwaysLogSuccesses()
+                    .alwaysLogFailures();
+
+                String pat;
+                try {
+                    pat = tokenProvider.getToken(userId, accessTokenRequest.getValidity(), accessTokenRequest.getScopes());
+                    rauditBuilder.success();
```

Comment:
It is really weird code. In case of success, the rauditx message is never issued. If this is the aim, we should create a message only in the catch statement. But I assume that more probably somebody omitted the call method `issue`.